lib/streamlit/__init__.py

```diff
@@ -580,6 +584,72 @@ def _transparent_write(*args):
     return args
 
 
+def _get_query_string():
+    """Return the query strings that is currently showing in the browser URL bar.
+    
+    Returns
+    -------
+    dict
+      The current query strings as a dict
+
+    Example
+    -------
+    >>> st.experimental.get_query_string()
+    .. output:: 
+       { "checkbox1" = "true" }
+
+    """
+    ctx = _get_report_ctx()
+    session_infos = Server.get_current()._session_info_by_id
+    _MAX_RETRIES = 3
+    _RETRY_WAIT_SECS = 0.2
+    for socket_handler, session_info in session_infos.items():
+        if session_info.session.enqueue == ctx.enqueue:
+            session = session_info.session
+            # Add retry in case query string message are not processed
+            for i in range(_MAX_RETRIES):
+                try:
+                    if (
+                        not hasattr(session, "query_string")
+                        or session.query_string is None
+                    ):
+                        raise StreamlitAPIException(
+                            "Error reading the current app's query strings from the browser. Please try again."
+                        )
+                    else:
+                        return session.query_string
+                except StreamlitAPIException as e:
+                    if i >= _MAX_RETRIES - 1:
+                        raise e
```

Comment:
If this is handling a race condition we should likely handle this differently.

Reply:
It is handling a race condition to be honest, any idea how should this be handled better? Appreciate any thoughts on this